### PR TITLE
NEW HBT_9 Heartbeat Test Case

### PR DESCRIPTION
**Pull request by RedaHossino**

### CODE REVIEW

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    that impacts incumbent activity.
+
+    successful heartbeat for the grants of two cbsds that don't impact the incumbent
+    and failed with responseCode 500 for the third one.
+    """
+    # STEP 1
+    # register three devices 
+    device_c = json.load(
+        open(os.path.join('testcases', 'testdata', 'device_c.json')))
+    device_d = json.load(
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: Device D is  Cat B and therefore we can't use single-step registration for it. We usually use A, C, E when we need three CBSDs and the category is unspecified.

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+        open(os.path.join('testcases', 'testdata', 'device_d.json')))
+    device_e = json.load(
+        open(os.path.join('testcases', 'testdata', 'device_e.json')))
+    devices = [device_d, device_e, device_c]
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: Why not use the same order as above?

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    for device in devices:
+        # inject FCC IDs
+        self._sas_admin.InjectFccId({'fccId': device['fccId']})
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: Please add a comment to later inject user ID. The code to add the function to the admin interface is in a pending PR from Google; hopefully we can submit that soon so that you can update the code instead of leaving a "TODO".

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    cbsd_ids.sort(key=lambda cbsdId: cbsdId)
+    del request, response    
+    # create and send grant requests
+    grant_c = json.load(
+        open(os.path.join('testcases', 'testdata', 'grant_0.json')))
+    grant_c['cbsdId'] = cbsd_ids[0]
+    grant_d = json.load(
+        open(os.path.join('testcases', 'testdata', 'grant_0.json')))
+    grant_d['cbsdId'] = cbsd_ids[1]
+    grant_d['operationParam']['operationFrequencyRange']['lowFrequency'] = 3560
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: The units here are Hz, not MHz.

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    grant_d['operationParam']['operationFrequencyRange']['highFrequency'] = 3570
+    grant_e = json.load(
+        open(os.path.join('testcases', 'testdata', 'grant_0.json')))
+    grant_e['cbsdId'] = cbsd_ids[2]
+    grant_e['operationParam']['operationFrequencyRange']['lowFrequency'] = 3560
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: This range isn't consistent with the latest test spec.

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py (hunk 4 above)
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: Why change this value for the second device? (Maybe it was in the old test spec but not the new one?)

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+            'grantId': grant_id,
+            'operationState': status
+        })
+    return {'heartbeatRequest': heartbeat_request}
+
+  def sendAndAssertSuccessfulHeartBeat(self, cbsd_ids, grant_ids, previousTransmitExpireTime):    
+    request =  self.getHeartBeatRequest(cbsd_ids, grant_ids, previousTransmitExpireTime)
+    response = self._sas.Heartbeat(request)['heartbeatResponse']
+    # check Heartbeat response 
+    self.assertEqual(len(response), 3)
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: You really can't hard-code this value in a general function.

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    for response_num, resp in enumerate(response):
+        self.assertEqual(resp['cbsdId'], cbsd_ids[response_num])
+        self.assertEqual(resp['grantId'], grant_ids[response_num])
+        self.assertEqual(resp['response']['responseCode'], 0)
+        grant_expire_time_1 = datetime.strptime(resp['grantExpireTime'],'%Y-%m-%dT%H:%M:%SZ')
+        transmit_expire_time_1 = datetime.strptime(resp['transmitExpireTime'],'%Y-%m-%dT%H:%M:%SZ')
+        self.assertLessEqual((transmit_expire_time_1 - datetime.utcnow()).total_seconds(), 240)
+        self.assertLessEqual(transmit_expire_time_1, grant_expire_time_1)
+    del request
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: There's no value to deleting the variable just before it goes out of scope anyway.

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+
+class HeartbeatTestcase(sas_testcase.SasTestCase):
+  def setUp(self):
+    self._sas, self._sas_admin = sas.GetTestingSas()
+    self._sas_admin.Reset()
+
+  def tearDown(self):
+    pass
+
+  def getHeartBeatRequest(self,cbsd_ids, grant_ids, previousTransmitExpireTime):
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: Do we really need all of this machinery? Since it's a very simple message, we usually just create the heartbeat request in the test case itself so that it's easier to read (and verify that it's doing the right thing). This is especially true for the HBT test cases where we often want to create the request in a very specific way. Is there a reason to do something else?

I'm also worried about the correctness of this function, but I'll hold off on those comments until we establish that it's a useful function to have.

(Same comment for the following function.)

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+    del request, response
+    # First Heartbeat
+    first_HB_response = self.sendAndAssertSuccessfulHeartBeat(cbsd_ids, grant_ids, [None, None, None])
+    transmitExpireTime = (first_HB_response[0]['transmitExpireTime'], first_HB_response[1]['transmitExpireTime'], first_HB_response[2]['transmitExpireTime'])
+    # STEP 2
+    # load and inject FSS data
+    fss_c = json.load(
+        open(os.path.join('testcases', 'testdata', 'fss_record_c.json')))
+    # Inject Incumbent Activity with Overlapping Frequency of CBSD
+    fss_c['deploymentParam']['operationParam']['operationFrequencyRange']['lowFrequency'] = \
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: Isn't this supposed to overlap with the third CBSD's grant, not the second?

Also it probably makes more sense to set this to 3650-4200 which is more consistent with the FSS data we expect (and you can do this in the file below).

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+                             grant_c['operationParam']['operationFrequencyRange']['lowFrequency']
+    fss_c['deploymentParam']['operationParam']['operationFrequencyRange']['highFrequency'] = \
+                             grant_c['operationParam']['operationFrequencyRange']['highFrequency']
+    
+    self._sas_admin.InjectFss({'record': fss_c})   
+    # load and inject GWPZ data
+    gwpz_c = json.load(
+        open(os.path.join('testcases', 'testdata', 'gwpz_record_c.json')))
+    # Inject Incumbent Activity with Overlapping Frequency of CBSD
+    gwpz_c['deploymentParam']['operationParam']['operationFrequencyRange']['lowFrequency'] = \
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: Let's choose a range that's more in line with the real data we expect for GWPZs (and you can do it in the file below; there's no need to override the value here).

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+                             grant_c['operationParam']['operationFrequencyRange']['lowFrequency']
+    gwpz_c['deploymentParam']['operationParam']['operationFrequencyRange']['highFrequency'] = \
+                             grant_c['operationParam']['operationFrequencyRange']['highFrequency']
+    
+    self._sas_admin.InjectWisp({'record': gwpz_c})
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: What about the 'zone' key? See https://github.com/Wireless-Innovation-Forum/Spectrum-Access-System/blob/master/src/harness/sas_interface.py#L288

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+    # Trigger daily activities and wait for it to get it complete
+    self.TriggerDailyActivitiesImmediatelyAndWaitUntilComplete()
+    # Send Heartbeat after triggering the IAP
+    request =  self.getHeartBeatRequest(cbsd_ids, grant_ids, transmitExpireTime)
+    response = self._sas.Heartbeat(request)['heartbeatResponse']
+    # CHECK
+    for response_num, resp in enumerate(response):
+        self.assertEqual(resp['cbsdId'], cbsd_ids[response_num])
+        self.assertEqual(resp['grantId'], grant_ids[response_num])
+        grant_expire_time_2 = datetime.strptime(resp['grantExpireTime'],'%Y-%m-%dT%H:%M:%SZ')
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: This variable name feels out of place. Why do you have "_2" at the end? (Same for other variable names.)

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+        transmit_expire_time_2 = datetime.strptime(resp['transmitExpireTime'],'%Y-%m-%dT%H:%M:%SZ')
+        self.assertLessEqual((transmit_expire_time_2 - datetime.utcnow()).total_seconds(), 240)
+        self.assertLessEqual(transmit_expire_time_2, grant_expire_time_2)
+        if cbsd_ids[response_num] == cbsd_ids[2]:
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: There's an easier way to write this condition, although normally we would split this block of code into two pieces to make it clearer:

// Checks for CBSDs 1, 2
// Check for CBSD 3 (since it's different)

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+            self.assertEqual(resp['response']['responseCode'], 500)
+            if request['heartbeatRequest'][response_num]['operationState'] == 'AUTHORIZED':
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: This condition has changed in the latest test spec.

File: src/harness/testcases/testdata/fss_record_c.json [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+  "id": "incumbent/ibfs/0",
+  "type": "FSS",
+  "deploymentParam": 
+    {
+      "installationParam": {
+        "latitude": 38.815291,
+        "longitude": -97.098805,
+        "height": 7.6,
+        "heightType": "AGL",
+        "indoorDeployment": false,
```
Comment (kate-harrison) [inline src/harness/testcases/testdata/fss_record_c.json]: Is this field applicable for an FSS record?

File: src/harness/testcases/testdata/fss_record_c.json [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+        "antennaGain": 16,
+        "antennaBeamwidth": 30,
+        "antennaAzimuth": 270,
+        "antennaDowntilt": -90
+      },
+      "operationParam": {
+        "maxEirp": 10.1,
```
Comment (kate-harrison) [inline src/harness/testcases/testdata/fss_record_c.json]: Does this make sense for an FSS record?

File: src/harness/testcases/testdata/fss_record_c.json [hunk 17]
```diff
@@ -0,0 +1,25 @@
+{
```
Comment (kate-harrison) [inline src/harness/testcases/testdata/fss_record_c.json]: I think you can just name this "fss_record.json" -- no need for the "_c" (same for the GWPZ file below).

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py (hunk 8 above)
Comment (RedaHossino) [reply src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: you are right, for the next function, because it is used only once in the test, but this function is used two times in the test, and i think it will be useful for other HB test cases so don't you think it is better to reuse it?
I think for this function it is also related to the new added machineryfor choosing the right status of the grant inside the HB request.
we need to verify that it is synchronized with the status in SAS" especially after the IAP which can take some time", other wise the SAS will send response 502, right?
so if you don't agree on this status machinery, we don't need to have this function.

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py (hunk 6 above)
Comment (RedaHossino) [reply src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: thank you this point is fixed, please verify this in the commits I added.

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py (hunk 7 above)
Comment (RedaHossino) [reply src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: you are right, thank you, it is done.

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py (hunk 1 above)
Comment (RedaHossino) [reply src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: I changes the used devices, to be a,c and e, thank you

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py (hunk 3 above)
Comment (RedaHossino) [reply src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: I modified the registration and grant requests, so I use assertRegisteredAndGranted

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py (hunk 4 above)
Comment (RedaHossino) [reply src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: you are right, thank you, it is done.

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py (hunk 5 above)
Comment (RedaHossino) [reply src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: I modified it, thank you.

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py (hunk 9 above)
Comment (RedaHossino) [reply src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: thank you, I modifies it to be on 3650-4200 MHz.

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py (hunk 10 above)
Comment (RedaHossino) [reply src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: I reset the range of the record to 3650-3700 MHz, which seems ok according to FCC data, right?

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py (hunk 13 above)
Comment (RedaHossino) [reply src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: fixed, thank you

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py (hunk 14 above)
Comment (RedaHossino) [reply src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: yes, you are right, I updated it.
thank you

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py (hunk 11 above)
Comment (RedaHossino) [reply src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: the issues of the comments related to the record files are done, please verify them in the latest change.
for the zone, I added it a polygon(rectangle) around the center, but we should modify it later according to realistic GWPZ shape data, is that ok ?
thank you

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+    # First Heartbeat
+    request =  self.getHeartBeatRequest(cbsd_ids, grant_ids, [None, None, None])
+    response = self._sas.Heartbeat(request)['heartbeatResponse']
+    # check Heartbeat response 
+    self.assertEqual(len(first_hb_response), len(grant_ids))
+    for response_num, resp in enumerate(first_hb_response):
+        self.assertEqual(resp['cbsdId'], cbsd_ids[response_num])
+        self.assertEqual(resp['grantId'], grant_ids[response_num])
+        self.assertEqual(resp['response']['responseCode'], 0)
+        grant_expire_time = datetime.strptime(resp['grantExpireTime'],\
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: I think we can avoid doing this and the following checks since they are not requested by the spec.

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+    # Send Heartbeat after triggering the IAP
+    request =  self.getHeartBeatRequest(cbsd_ids, grant_ids,\
+                                        first_hb_transmit_expire_time)
+    response = self._sas.Heartbeat(request)['heartbeatResponse']
+    # CHECK
+    # Check first two CBSDs with a successful HB response
+    for response_num, resp in enumerate(response[:2]):
+        self.assertEqual(resp['cbsdId'], cbsd_ids[response_num])
+        self.assertEqual(resp['grantId'], grant_ids[response_num])
+        grant_expire_time = datetime.strptime(resp['grantExpireTime'],\
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: We don't expect grantExpireTime to be set in the HeartbeatResponse (except if grantRenew==True, which it does not in this case). This value should be obtained from the GrantResponse.

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py [hunk 20]
```diff
@@ -0,0 +1,135 @@
...
+                                                 '%Y-%m-%dT%H:%M:%SZ')
+        self.assertLessEqual((transmit_expire_time - \
+                              datetime.utcnow()).total_seconds(), 240)
+        self.assertLessEqual(transmit_expire_time, grant_expire_time)
+        self.assertEqual(resp['response']['responseCode'], 0)
+    # Check the third  CBSD with unsuccessful HB response of code 500
+    self.assertEqual(response[2]['cbsdId'], cbsd_ids[2])
+    self.assertEqual(response[2]['grantId'], grant_ids[2])
+    self.assertEqual(response[2]['response']['responseCode'], 500)
+    self.assertLessEqual((datetime.utcnow() - \
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: You can compare the times directly without having to subtract.

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py [hunk 21]
```diff
@@ -0,0 +1,135 @@
...
+                                   first_hb_response[2]['transmitExpireTime'])
+    # STEP 2
+    # load and inject FSS data with Overlapping Frequency of CBSD
+    fss_e = json.load(
+        open(os.path.join('testcases', 'testdata', 'fss_record.json')))
+    self._sas_admin.InjectFss({'record': fss_e})   
+    # load and inject GWPZ data with Overlapping Frequency of CBSD 
+    gwpz_e = json.load(
+        open(os.path.join('testcases', 'testdata', 'gwpz_record.json')))  
+    self._sas_admin.InjectWisp({'record': gwpz_c})
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: Given the way you defined gwpz_record.json, you don't need 'record' here.

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py (hunk 8 above)
Comment (kate-harrison) [reply src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: Aha, now I understand your concern. I think the machinery is unnecessary during the first heartbeat request (we know that it should always be GRANTED). For the second heartbeat request, I think we can safely use GRANTED as well -- even if the SAS thinks the Grant is AUTHORIZED, it should not give an error. What do you think?

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py (hunk 10 above)
Comment (kate-harrison) [reply src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: Agreed -- thanks!

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py (hunk 11 above)
Comment (kate-harrison) [reply src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: This sounds good!

File: src/harness/testcases/testdata/fss_record_c.json (hunk 17 above)
Comment (kate-harrison) [reply src/harness/testcases/testdata/fss_record_c.json]: Sorry, my original suggestion was slightly flawed -- I should have suggested "fss_record_0.json" (so that we can create more in the future, if needed; this will match what we've done for similar data files).

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py (hunk 8 above)
Comment (RedaHossino) [reply src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: yeah, you are right, I thought the CBSD should always send in HB the status which exists in the SAS otherwise the SAS will send error(502), but I think you are right if the CBSD sends granted status in the HB request and its status on the SAS is Authorized it is safe, and there should be no error.
We don't need this machinery then, thank you.

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py (hunk 20 above)
Comment (RedaHossino) [reply src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: thank you for the review, I updated the test according to the requested changes.

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+    grant_c = json.load(
+        open(os.path.join('testcases', 'testdata', 'grant_0.json')))
+    grant_c['cbsdId'] = cbsd_ids[1]
+    grant_e = json.load(
+        open(os.path.join('testcases', 'testdata', 'grant_0.json')))
+    grant_e['cbsdId'] = cbsd_ids[2]
+    grant_e['operationParam']['operationFrequencyRange']\
+                             ['lowFrequency'] = 3650000000
+    grant_e['operationParam']['operationFrequencyRange']\
+                             ['highFrequency'] = 3700000000
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: Why did you change this to 3660000000? What you had before matches the spec (although this version does as well).

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py [hunk 23]
```diff
@@ -0,0 +1,131 @@
...
+    request = {'grantRequest': [grant_a, grant_c, grant_e]}
+    # send grant request
+    response = self._sas.Grant(request)['grantResponse']
+    grant_ids = []
+    grant_expire_times = []
+    for response_num, resp in enumerate(response):
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: Please assert that the responseCode for each is 0 (it's not requested by the spec but it could help a lot during debugging and doesn't cause any problems otherwise).

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py [hunk 24]
```diff
@@ -0,0 +1,131 @@
...
+    response = self._sas.Heartbeat({'heartbeatRequest': request})['heartbeatResponse']
+    del request
+    # CHECK
+    # Check first two CBSDs with a successful HB response
+    for response_num, resp in enumerate(response[:2]):
+        self.assertEqual(resp['cbsdId'], cbsd_ids[response_num])
+        self.assertEqual(resp['grantId'], grant_ids[response_num])
+        self.assertEqual(resp['response']['responseCode'], 0)
+        transmit_expire_time = datetime.strptime(resp['transmitExpireTime'],\
+                                                 '%Y-%m-%dT%H:%M:%SZ')
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: Please also check that transmit_expire_time >= Now().

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py [hunk 25]
```diff
@@ -0,0 +1,133 @@
...
+    # First Heartbeat
+    request = []
+    for cbsd_id, grant_id in zip(cbsd_ids, grant_ids):
+        request.append({
+            'cbsdId': cbsd_id,
+            'grantId': grant_id,
+            'operationState': 'GRANTED'
+        })
+    
+    response = self._sas.Heartbeat({'heartbeatRequest': request})['heartbeatResponse']
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: Please assert that the HB response code for each is 0.

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py [hunk 26]
```diff
@@ -0,0 +1,133 @@
...
+    
+    del request, response
+    # STEP 2
+    # load and inject FSS data with Overlapping Frequency of CBSD
+    fss_e = json.load(
+        open(os.path.join('testcases', 'testdata', 'fss_record_0.json')))
+    self._sas_admin.InjectFss({'record': fss_e})   
+    # load and inject GWPZ data with Overlapping Frequency of CBSD
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: Technically this should be GWBL data. I think you can just change "gwpz" to "gwbl" here and in the name of the file to avoid confusion.

File: src/harness/testcases/testdata/gwpz_record_0.json [hunk 27]
```diff
@@ -0,0 +1,58 @@
...
+      },
+      "operationParam": {
+        "operationFrequencyRange": {
+          "lowFrequency": 3650000000.0,
+          "highFrequency": 3700000000.0
+        }
+      }
+    }
+},
+"zone":{
```
Comment (kate-harrison) [inline src/harness/testcases/testdata/gwpz_record_0.json]: I think you can actually leave out the zone here since the intent is to inject a GWBL, not a GWPZ (see the documentation [here](https://docs.google.com/document/d/19YbeQsSzm5kCIOeVzsF8piqI0Lizy3p345aiVGf-Hlg/edit?pli=1#heading=h.rz2qbup2aut9) which someone else helpfully reminded me of on Tuesday; sorry for the back-and-forth on this, it's completely my fault).

File: src/harness/testcases/WINNF_FT_S_HBT_testcase.py (hunk 25 above)
Comment (redtechnologies) [reply src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: Thank you for the very good review, I addressed them, please feel free to look at them.

Comment (RedaHossino) [reply src/harness/testcases/WINNF_FT_S_HBT_testcase.py]: Thank you for the very good review, I addressed them, please feel free to look at them.

File: src/harness/testcases/testdata/fss_record_0.json [hunk 28]
```diff
@@ -0,0 +1,23 @@
+{
+  "id": "incumbent/ibfs/0",
+  "type": "FSS",
+  "deploymentParam":
```
Comment (kate-harrison) [inline src/harness/testcases/testdata/fss_record_0.json]: I think Takashi's comment from SIQ.2 applies here as well: deploymentParam should be an array of DeploymentParam objects rather than a single object (same for the GWBL below). I'll try to make this clearer in the testing API doc -- I think it's my fault that it's not very clear right now.
